libsolidity/lsp/DocumentHoverHandler.cpp

```diff
@@ -0,0 +1,126 @@
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/lsp/DocumentHoverHandler.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <fmt/format.h>
+
+namespace solidity::lsp
+{
+
+using namespace std;
+
+using namespace solidity::lsp;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace
+{
+
+struct MarkdownBuilder
+{
+	std::stringstream result;
+
+	MarkdownBuilder& code(std::string const& _code)
+	{
+		auto constexpr SolidityLanguageId = "solidity";
+		result << "```" << SolidityLanguageId << '\n' << _code << "\n```\n\n";
+		return *this;
+	}
+
+	MarkdownBuilder& paragraph(std::string const& _text)
+	{
+		if (!_text.empty())
+		{
+			result << _text << '\n';
+			if (_text.back() != '\n') // We want double-LF to ensure constructing a paragraph.
+				result << '\n';
+		}
+		return *this;
+	}
+};
+
+}
+
+void DocumentHoverHandler::operator()(MessageID _id, Json::Value const& _args)
+{
+	auto const [sourceUnitName, lineColumn] = HandlerBase(*this).extractSourceUnitNameAndLineColumn(_args);
+	auto const [sourceNode, sourceOffset] = m_server.astNodeAndOffsetAtSourceLocation(sourceUnitName, lineColumn);
+
+	MarkdownBuilder markdown{};
+	//markdown.paragraph(fmt::format("This is: {}", typeid(*sourceNode).name()));
+	auto rangeToHighlight = toRange(sourceNode->location());
+
+	// Try getting the type definition of the underlying AST node, if available.
+	if (auto const* expression = dynamic_cast<Expression const*>(sourceNode))
+	{
+		if (auto const* declaration = ASTNode::referencedDeclaration(*expression))
+			if (declaration->type())
+				markdown.code(declaration->type()->toString(false));
+	}
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(sourceNode))
+	{
+		if (declaration->type())
+			markdown.code(declaration->type()->toString(false));
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(sourceNode))
+	{
+		for (size_t i = 0; i < identifierPath->path().size(); ++i)
+		{
+			if (!identifierPath->pathLocations()[i].containsOffset(sourceOffset))
+				continue;
+
+			rangeToHighlight = toRange(identifierPath->pathLocations()[i]);
+
+			if (i < identifierPath->annotation().pathDeclarations.size())
+			{
+				Declaration const* declaration = identifierPath->annotation().pathDeclarations[i];
+				if (declaration && declaration->type())
+					markdown.code(declaration->type()->toString(false));
+				if (auto const* structurallyDocumented = dynamic_cast<StructurallyDocumented const*>(declaration))
+					if (structurallyDocumented->documentation()->text())
+						markdown.paragraph(*structurallyDocumented->documentation()->text());
+			}
+		}
+	}
+
+	// If this AST node contains documentation itself, append it.
+	if (auto const* documented = dynamic_cast<StructurallyDocumented const*>(sourceNode))
+	{
+		if (documented->documentation())
+			markdown.paragraph(*documented->documentation()->text());
+	}
```

Comment:
Can `IdentifierPath` be `StructurallyDocumented`? If it can, you'll display the documentaion twice for it.

Reply:
`IdentifierPath` is NOT `StructurallyDocumented`, it's just an `ASTNode` without any other base classes. :)